File: python/스택큐/프로세스.py

""" Level 2. 프로세스 """

def solution(priorities, location):
    answer = 0

    P = [(pr, idx) for idx, pr in enumerate(priorities)]  
    N = len(priorities) - 1  

    # while len(P) > 1:
    while answer < N:
        
        temp = P.pop(0)

        if any(temp[0] < p[0] for p in P):
            P.append(temp)
            continue
        
        answer += 1

        if temp[1] == location:
            break
    
    else:
        answer += 1

    return answer

print(solution([2, 1, 3, 2], 2))            # 1
print(solution([1, 1, 9, 1, 1, 1], 0))      # 5

print(solution([1, 1, 1, 2], 2))            # 4

File: Java/src/자료구조/프로세스.java
package 자료구조;

import java.util.*;

public class 프로세스 {
    
    public static void main(String[] args) {

        int[] p1 = {2, 1, 3, 2};
        int l1 = 2;
        System.out.println(solution(p1, l1));

        int[] p2 = {1, 1, 9, 1, 1, 1};
        int l2 = 0;
        System.out.println(solution(p2, l2));
    }

    public static int solution(int[] priorities, int location) {

        Queue<int[]> q = new LinkedList<>();

        for (int i = 0; i < priorities.length; i++) {
            int[] temp = {i, priorities[i]};
            q.add(temp);
        }

        Integer[] priority = Arrays.stream(priorities).boxed().toArray(Integer[]::new);
        Arrays.sort(priority, Collections.reverseOrder());

        int pdx = 0;
        int answer = 0;

        while (!q.isEmpty()) {

            int[] temp = q.poll();

            if (temp[1] == priority[pdx]) {
                pdx += 1;
                answer += 1;
                if (temp[0] == location) return answer;
            }

            q.add(temp);
        }
        
        return answer;
    }

}
